### PR TITLE
Database helpers

### PR DESCRIPTION
**Pull request by KnightEther**
Add DatabaseElement that give information of synchronization with database, trigger propertyChange event when setting properties
Add CollectionManager
Both inform if tasks are running with their implementation of BusyElement

User and Project implement DatabaseElement
The user search use now the CollectionManager

### CODE REVIEW

File: app/src/androidTest/java/ch/epfl/sweng/thegramginers/SearchUserTests.java [hunk 1]
```diff
@@ -68,26 +70,18 @@ public void testRecyclerViewIsDisplayed() {
...
     @Test
     public void testSwipeDownToRefresh(){
-        final List<User> newUsers = new ArrayList<>();
         User user1 = new User("1");
         User user2 = new User("2");
-        newUsers.add(user1);
-        newUsers.add(user2);
-        usersFragment.updateUsers(newUsers);
+        TestUtils.waitElementToBeSynchronized(user1);
+        TestUtils.waitElementToBeSynchronized(user2);
```
Comment (alemannosara) [inline app/src/androidTest/java/ch/epfl/sweng/thegramginers/SearchUserTests.java]: It is fine that we can use the same function for each element. Great job! 😍

File: app/src/androidTest/java/ch/epfl/sweng/thegramginers/TestUtils.java [hunk 2]
```diff
@@ -45,39 +47,43 @@ protected Void doInBackground(Void... voids) {
...
-        AsyncTask<Void, Void, Void> waitReadyUser = new AsyncTask<Void,Void,Void>(){
+    public static void waitElementToBeSynchronized(final DatabaseElement element) {
+        AsyncTask<Void, Void, Void> waitSynchronized = new AsyncTask<Void,Void,Void>(){
             @Override
             protected Void doInBackground(Void... voids) {
-                while(!currentUser.getIsReady());
+                while(!element.isSynchronized());
                 return null;
             }
         };
```
Comment (alemannosara) [inline app/src/androidTest/java/ch/epfl/sweng/thegramginers/TestUtils.java]: This means 
• synchronized : we are done getting it from db 
• idle : the user is not setting anything in the database or doing something that makes him « busy » 
Is that correct?

File: app/src/main/java/ch/epfl/sweng/thegramginers/DatabaseHelper/DatabaseElement.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        return properties.get(propertyName);
+    }
+    //endregion
+
+    //region Setters
+    //Update property only in local
+    public void setProperty(String propertyName, Object newValue){
+        setProperty(propertyName, newValue, false);
+    }
+
```
Comment (alemannosara) [inline app/src/main/java/ch/epfl/sweng/thegramginers/DatabaseHelper/DatabaseElement.java]: Then I can reset all the properties and update when I am done? That is very cool and efficient 💪🏼

File: app/src/main/java/ch/epfl/sweng/thegramginers/ProfileEditionActivity.java [hunk 4]
```diff
@@ -69,30 +69,30 @@ public void saveButtonOnClick(View view) {
...
         }else{
-            currentUser.setWebsite(website);
+            currentUser.updateWebsite(website);
         }
 
-        currentUser.setCountry(country);
+        currentUser.updateCountry(country);
 
-        currentUser.setCity(city);
+        currentUser.updateCity(city);
```
Comment (alemannosara) [inline app/src/main/java/ch/epfl/sweng/thegramginers/ProfileEditionActivity.java]: Why don’t you set everything and then update in the db?

File: app/src/androidTest/java/ch/epfl/sweng/thegramginers/TestUtils.java (hunk 2 above)
Comment (KnightEther) [reply app/src/androidTest/java/ch/epfl/sweng/thegramginers/TestUtils.java]: Yes that's correct.
We have to ways of setting a database object : set or update which will modify on local only or both
If we modify on local we loose synchronization with the db and we get it back when we push our new properties.

File: app/src/main/java/ch/epfl/sweng/thegramginers/ProfileEditionActivity.java (hunk 4 above)
Comment (KnightEther) [reply app/src/main/java/ch/epfl/sweng/thegramginers/ProfileEditionActivity.java]: I want to only set them when they get indeed modify by the database such that we don't think it is done and it is not. But it's true that with the new implementation I may think back.

File: app/src/main/java/ch/epfl/sweng/thegramginers/DatabaseHelper/DatabaseElement.java (hunk 3 above)
Comment (KnightEther) [reply app/src/main/java/ch/epfl/sweng/thegramginers/DatabaseHelper/DatabaseElement.java]: Yes indeed

**Pull request merged**